### ISSUE TITLE
Poor performance from BigQuery API when retrieving a large volume of data

### ISSUE THREAD

**Issue by BommieRobear** [googleapis/google-cloud-dotnet#8231]
- OS: Windows 10
  - .NET version: 5.0
  - Package name and version: Google.Cloud.BigQuery.V2

#### Steps to reproduce

  1. Create a query in BigQuery returning with 4+ million rows, exceeding 2GB of data.
  2. Execute the query to using `.ExecuteQueryAsync(query, null, new QueryOptions { UseQueryCache = false });`
  3. Iterate over the dataset using `.GetRowsAsync()`.

#### Expected Results
Data is returned from the API in < 5 seconds.

#### Actual Results
Data is returned from the API in 6+ minutes

#### Workaround
Prepend the query with a `CREATE TABLE [...] AS` and retrieve the data using the `Google.Cloud.BigQuery.Storage.V2`.

#### Description
In our use-case, we're executing a query that unions two queries that have a single join each. The query executes in BigQuery directly in ~35 seconds. This same query, executed using the procedure outlined above, takes in excess of 6 minutes to return data. By writing to a temporary table and streaming that via the storage API, we were able to get the data in ~40 seconds. This is the behavior we expected from the `Google.Cloud.BigQuery.V2` API.

**Comment by jskeet** [comment]
I think if the query is executing in the BigQuery UI in 35 seconds (and note that that *won't* be displaying all the data - just the first page of it) then an expectation of getting all the data back to another machine in less than 5 seconds is unrealistic.

There are two stages here:

- Performing the query, which I believe always stores the results *somewhere* whether the destination is explicitly specified or not
- Retrieving the data over the network

You should be able to observe the timing of the different stages: the task returned by `ExecuteQueryAsync` won't complete until the query job itself is complete (so the data is ready to *start* transferring), and then you can start iterating over the rows. You can fetch the `Job` created for the query, and look at its `JobStatistics`, which will show the creation and completion times. During the first stage I'd expect you to see almost no network traffic: just enough to start the job and poll for completion.

Note that instead of specifying `CREATE TABLE [...] AS` in the query, you can specify the `DestinationTable` property in `QueryOptions` (which is the type of the optional parameter `queryOptions` in `ExecuteQueryAsync`) - you may find that simpler to work with than modifying the query.

I suspect that you'll find the first stage takes ~30-35 seconds, and the remainder is taken transferring the data from the new table. This is effectively a limitation of the API being used: each value is serialized as a separate JSON object. This is inefficient in terms of:

- Serialization to JSON at the server side
- Wire transfer (which is compressed by default, but still not ideal)
- Deserialization from JSON at the client side

There's nothing the client library can do to make that faster though (while using the existing API): it's the nature of the underlying API. That's precisely why the BigQuery Storage Read API exists: to transfer data more efficiently, both in terms of the wire format and the ability to parallelize operations.

In conclusion, I don't think the Google.Cloud.BigQuery.V2 library (which is *only* intended to be a library for the main BigQuery JSON API, *not* the BigQuery Storage Read API) is doing anything wrong here. It's possible that it could be *slightly* faster if it could issue requests for the next page of data in advance while deserializing the current page, but I strongly suspect that won't make *much* difference, and would introduce very significant complexity. Basically, if you're trying to transfer all the data from a very large table (whether that's the result of a query or not), the BigQuery Storage Read API will be a more efficient approach.

**Comment by BommieRobear** [comment]
> then an expectation of getting all the data back to another machine in less than 5 seconds is unrealistic.

This was my mistake as I should have been more clear: the expectation is that the data _transfer_ following the query execution would take less than 5 seconds, as it does with the storage data API. The execution end-to-end using the workaround provides us the data in ~40 seconds, which is the expected behavior of the BigQuery API.

> you can specify the DestinationTable property in QueryOptions 

We took this approach initially but replaced it because we also wanted to specify an expiration date explicitly in the event that the `finally` block failed to delete the temporary table.

> I don't think the Google.Cloud.BigQuery.V2 library (which is only intended to be a library for the main BigQuery JSON API, not the BigQuery Storage Read API) is doing anything wrong here.

Is there a different API designed for specifically for query execution we should be using?

**Comment by jskeet** [comment]
> We took this approach initially but replaced it because we also wanted to specify an expiration date explicitly in the event that the finally block failed to delete the temporary table.

Right. I don't know whether there's any way of setting the expiration directly in the API in the way that we specify the destination table, but I could look into that next week if you'd like.

> Is there a different API optimized for query execution we should be using?

No, I think you're basically doing the right thing: start the job with the "main" API, and then read the results with the BigQuery Storage API. I don't know of any individual API that combines the "best of both worlds" I'm afraid. (Of course, if the BigQuery team releases such an API, we'll create a .NET library to make that easily available.)

**Comment by jskeet** [comment]
Closing this as I don't believe there's any further work to do here, while the APIs are as they are.
